### PR TITLE
Add capabilites to change application theme

### LINKED ISSUE #11: Customize the way the application looks
As an Artsy User it allows me to customize the looks of the application to blend it into my workflow better.

### PR DESCRIPTION
**Pull request by lue97**
Fixes #11 

AB3 offers no control over the colors of the UI. This may lead to a bad
user experience as the application may not appear "cohesive" with the
rest of the system.

Including a method of allowing users to change colors of the GUI allows
more control over how the application looks.

Lets create a system that allows the user to customize the UI.

### CODE REVIEW

File: src/main/java/seedu/address/MainApp.java [hunk 1]
```diff
@@ -120,7 +151,7 @@ protected Config initConfig(Path configFilePath) {
             initializedConfig = configOptional.orElse(new Config());
         } catch (DataConversionException e) {
             logger.warning("Config file at " + configFilePathUsed + " is not in the correct format. "
-                    + "Using default config properties");
+                + "Using default config properties");
```
Comment (chesterhow) [inline src/main/java/seedu/address/MainApp.java]: Indentation should be 8 spaces

File: src/main/java/seedu/address/MainApp.java [hunk 2]
```diff
@@ -148,7 +179,7 @@ protected UserPrefs initPrefs(UserPrefsStorage storage) {
             initializedPrefs = prefsOptional.orElse(new UserPrefs());
         } catch (DataConversionException e) {
             logger.warning("UserPrefs file at " + prefsFilePath + " is not in the correct format. "
-                    + "Using default user prefs");
+                + "Using default user prefs");
```
Comment (chesterhow) [inline src/main/java/seedu/address/MainApp.java]: Same here indentation issue

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -144,8 +144,8 @@ public boolean equals(Object obj) {
         // state check
         ModelManager other = (ModelManager) obj;
         return addressBook.equals(other.addressBook)
-                && userPrefs.equals(other.userPrefs)
-                && filteredPersons.equals(other.filteredPersons);
+            && userPrefs.equals(other.userPrefs)
+            && filteredPersons.equals(other.filteredPersons);
```
Comment (chesterhow) [inline src/main/java/seedu/address/model/ModelManager.java]: Indentation issue as well

File: src/main/java/seedu/address/ui/ThemeManager.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    private static String cssCacheUri = null;
+    /**
+     * The mainScene property from a MainWindow instance.
+     */
+    private static Scene scene = null;
+
+    static {
+        InputStream templateStream = MainApp.class.getResourceAsStream("/view/Template.css");
+        CSS_TEMPLATE = new BufferedReader(new InputStreamReader(templateStream))
+            .lines().collect(Collectors.joining("\n"));
```
Comment (chesterhow) [inline src/main/java/seedu/address/ui/ThemeManager.java]: Indentation issue as well

File: src/main/java/seedu/address/ui/ThemeManager.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+     * @return cssTemplate with colors assigned.
+     */
+    private static String generateCssFromTheme(Theme theme) {
+        String cssString = ThemeManager.CSS_TEMPLATE;
+        cssString = cssString
+            .replaceAll("\\$foreground", theme.foreground)
+            .replaceAll("\\$background", theme.background);
+        for (int i = 0; i < 16; i++) {
+            cssString = cssString
+                .replaceAll("\\$c" + Integer.toHexString(i), theme.color[i]);
```
Comment (chesterhow) [inline src/main/java/seedu/address/ui/ThemeManager.java]: Same here for lines 132, 133 and 136

File: src/main/resources/view/Template.css [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    -fx-text-fill: $foreground;
+    -fx-opacity: 0.9;
+}
+
+.label-bright {
+    -fx-font-size: 11pt;
+    -fx-font-family: "Segoe UI Semibold";
+    -fx-text-fill: $foreground;
+    -fx-opacity: 1;
+    -fx-opacity: 1;
```
Comment (chesterhow) [inline src/main/resources/view/Template.css]: Duplicate opacity styles

File: src/test/data/ThemeTest/default.template.css [hunk 7]
```diff
@@ -1,26 +1,27 @@
...
+    -fx-text-fill: #f8f8f2;
     -fx-opacity: 0.9;
 }
 
 .label-bright {
     -fx-font-size: 11pt;
     -fx-font-family: "Segoe UI Semibold";
-    -fx-text-fill: white;
+    -fx-text-fill: #f8f8f2;
+    -fx-opacity: 1;
```
Comment (chesterhow) [inline src/test/data/ThemeTest/default.template.css:17]: Duplicate opacity styles here as well

File: src/test/java/seedu/address/ui/ThemeManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        ThemeManager.init();
+        assertEquals(ThemeManager.getTheme(), ThemeFactory.getDefaultTheme());
+    }
+
+    @Test
+    public void themeManager_setTheme_success() {
+        ThemeManager.setTheme(
+            ThemeFactory.getDefaultTheme(), "src/test/data/ThemeTest/valid.theme.json");
+        String defaultCss = assertDoesNotThrow(() -> Files
+            .readString(Path.of("src/test/data/ThemeTest/default.template.css")));
```
Comment (chesterhow) [inline src/test/java/seedu/address/ui/ThemeManagerTest.java]: Indentation issues on line 22 and 24

File: src/main/java/seedu/address/ui/ThemeFactory.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    private static final String[] DEFAULT_COLOR;
+
+    static {
+        DEFAULT_COLOR = new String[] {
+            "#272822", "#f92672", "#a6e22e", "#f4bf75",
+            "#66d9ef", "#ae81ff", "#a1efe4", "#f8f8f2",
+            "#75715e", "#f92672", "#a6e22e", "#f4bf75",
+            "#66d9ef", "#ae81ff", "#a1efe4", "#f9f8f5"
+        };
+    }
```
Comment (chesterhow) [inline src/main/java/seedu/address/ui/ThemeFactory.java:25]: Not too sure about this, but should this be `private static final`?

**Review by chesterhow — APPROVED** [review]
LGTM, just some minor formatting issues. Good to merge once you've amended it!

**Review by Assyarul — APPROVED** [review]
LGTM!

**Pull request merged**